sklearn/cluster/hierarchical.py

```diff
@@ -788,10 +804,14 @@ def fit(self, X, y=None):
         X = check_array(X, ensure_min_samples=2, estimator=self)
         memory = check_memory(self.memory)
 
-        if self.n_clusters <= 0:
+        if self.n_clusters is not None and self.n_clusters <= 0:
             raise ValueError("n_clusters should be an integer greater than 0."
                              " %s was provided." % str(self.n_clusters))
 
+        if self.n_clusters is None and self.distance_threshold is None:
+            raise ValueError("n_clusters and distance_threshold cannot be "
+                             "both None.")
+
```

Comment:
We also need to check that:

- n_clusters is None if distance_threshold is not None
- distance_threshold is None if n_clusters is not None
- compute_full_tree is True if distance_threshold is not None

along with the associated tests.

Please also update n_clusters docstring to

>    n_clusters : int or None, default=2

Reply:
I think the decision was that the `n_clusters` is ignored if `distance_threshold` is given. So it doesn't have to explicitly set as `None`, and that's why the first two checks are not here in the code. I don't have a strong position on either case.

> compute_full_tree is True if distance_threshold is not None

Good point.

Reply:
> I think the decision was that the n_clusters is ignored if distance_threshold is given

Ok. How about not forcing it to be None, but still forcing it to have its default value (2, I think)?

Reply:
And what would we do if the default value is explicitly given by the user? Do we want to detect that as well?

Reply:
well you can't

Reply:
That's why I think forcing it to none is a better idea.
It seems to me that we usually raise error when 2 exclusive attributes are used simultaneously.

Reply:
Ok, done.